addons/mail/models/mail_thread.py

```diff
@@ -2008,112 +2022,226 @@ def message_post(self, *,
-
-        # Create the composer
-        composer = self.env['mail.compose.message'].with_context(
-            active_id=res_id,
-            active_ids=res_ids,
-            active_model=kwargs.get('model', self._name),
-            default_composition_mode=kwargs['composition_mode'],
-            default_email_layout_xmlid=email_layout_xmlid,
-            default_model=kwargs.get('model', self._name),
-            default_res_id=res_id,
-            default_template_id=template_id,
-        ).create(kwargs)
-        # Simulate the onchange (like trigger in form the view) only
-        # when having a template in single-email mode
-        if template_id:
-            update_values = composer._onchange_template_id(template_id, kwargs['composition_mode'], self._name, res_id)['value']
-            composer.write(update_values)
-        return composer._action_send_mail(auto_commit=auto_commit)
+        view_ref_valid = view_ref.id if isinstance(view_ref, models.BaseModel) else view_ref
+        if (not view_ref_valid and not template_id) or (view_ref_valid and template_id):
+            raise ValueError(_('Should give either template, either view'))
+
+        # preliminary value safety check
+        self._raise_for_invalid_fields([kwargs], ('body', 'composition_mode', 'model', 'res_id', 'values'))
+
+        # with a view, render bodies in batch (template is managed by composer)
+        bodies = self.env['mail.render.mixin']._render_template_qweb_view(
+            view_ref_valid,
+            self._name,
+            self.ids,
+            add_context=render_values,
+        ) if view_ref_valid else {}
+
+        # Prepare composer values for creation
+        subtype_id = kwargs.pop('subtype_id', False) or self.env['ir.model.data']._xmlid_to_res_id('mail.mt_note')
+        composer_values = {
+            'composition_mode': 'comment',
+            'message_type': message_type,
+            'subtype_id': subtype_id,
+        }
+        composer_values.update(**kwargs)
+
+        messages = self.env['mail.message']
+        for record in self:
+            if template_id:
+                composer = self.env['mail.compose.message'].with_context(
+                    active_ids=record.ids,  # still required for mass mail mode
+                    default_composition_mode='comment',
+                    default_model=self._name,
+                    default_res_id=record.id,
+                    default_template_id=template_id,
+                ).create(composer_values)
+
+                # Simulate the onchange (like trigger in form the view) only
+                # when having a template in single-email mode
+                if template_id:
+                    update_values = composer._onchange_template_id(
+                        template_id,
+                        composer_values['composition_mode'],
+                        self._name,
+                        self.ids[0] if self.ids else 0
```

Comment:
shouldn't we give `record.id` instead ? (it's impossible for self.ids to be empty here, because we iterate over self above)

Reply:
Yep, bad copy-paste-move that can be cleaned :)